ducktape/tests/action.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2015 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Provides pass/fail/aux-results reporting for specific test actions without
+# necessarily affecting the entire test outcome.
+#
+
+import traceback
+import inspect
+import time
+
+class TestAction (object):
+    # Accumulated test actions in various forms.
+    actions_root = list()
+    actions = list()
+    actions_by_fullname = dict()
+
+    def __init__ (self, name, instance=None, desc=None, parent=None, logger=None):
+        """
+            name     := Appended to parent's name to create unique action identifier
+            instance := Instance to test: allows multiple actions with the same name
+            desc     := Human readable description of this action
+            parent   := Parent action, either a TestAction object or a fully qualified
+                        action name (as returned by TestAction.fullname).
+            logger   := Logger to use. Only set this on your top-level action, it will be
+                        inherited to all sub-actions.
+        """
+        
+        super(TestAction, self).__init__()
+        self.name       = name
+        self.instance   = instance
+        self.desc       = desc
+        self.logger     = logger
+        self.start_time = time.time()
+        self.duration   = 999.999
+        # Resolve parent
+        if type(parent) == str:
+            if parent in actions:
+                parent = actions[parent]
+            else:
+                parent = None
+        self.parent   = parent
+
+        prefix=''
+        if self.parent:
+            # Add us to parent
+            self.parent.subs.append(self)
+            # Use parent's logger
+            if not self.logger and self.parent.logger:
+                self.logger = self.parent.logger
+            # Prefix our name with parent's name
+            prefix = self.parent.fullname + '.'
+
+
+        # If there is no description but the calling function has a doc string, use that,
+        # unless the parent action has the same one (a bit repetative).
+        # FIXME: Doesnt seem to work for class functions
```

Comment:
I assume that using the calling stack frame has some motivation, but I'm unclear why so right now I'd suggest just dropping this bit of code. Is there any reason not to use the parent as a fallback and nothing if there is no parent or the parent has no doc (and perhaps continue upward as necessary)?